proxy/http/HttpSessionManager.cc

```diff
@@ -64,41 +64,88 @@ ServerSessionPool::purge()
+      Debug("http_ss", "validate_host_sni host=%*.s, sni=%s", len, req_host, session_sni);
+    }
+  }
+  return retval;
 }
 
 bool
 ServerSessionPool::validate_sni(HttpSM *sm, NetVConnection *netvc)
 {
-  // TS-4468: If the connection matches, make sure the SNI server
-  // name (if present) matches the request hostname
-  int len              = 0;
-  const char *req_host = sm->t_state.hdr_info.server_request.host_get(&len);
-  // The sni_servername of the connection was set on HttpSM::do_http_server_open
-  // by fetching the hostname from the server request.  So the connection should only
-  // be reused if the hostname in the new request is the same as the host name in the
-  // original request
-  const char *session_sni = netvc->options.sni_servername;
-
-  return ((sm->t_state.scheme != URL_WKSIDX_HTTPS) || !session_sni || strncasecmp(session_sni, req_host, len) == 0);
+  bool retval = true;
+  // Verify that the sni name on this connection would match the sni we would have use to create
+  // a new connection.
+  //
+  if (sm->t_state.scheme == URL_WKSIDX_HTTPS) {
+    const char *session_sni       = netvc->options.sni_servername;
+    std::string_view proposed_sni = sm->get_outbound_sni();
+    Debug("http_ss", "validate_sni proposed_sni=%s, sni=%s", proposed_sni.data(), session_sni);
+    if (!session_sni || proposed_sni.length() == 0) {
+      retval = session_sni == nullptr && proposed_sni.length() == 0;
+    } else {
+      retval = proposed_sni.compare(session_sni) == 0;
+    }
+  }
+  return retval;
+}
+
+bool
+ServerSessionPool::validate_cert(HttpSM *sm, NetVConnection *netvc)
+{
+  bool retval = true;
+  // Verify that the cert file associated this connection would match the cert file we would have use to create
+  // a new connection.
+  //
+  if (sm->t_state.scheme == URL_WKSIDX_HTTPS) {
+    const char *session_cert       = netvc->options.ssl_client_cert_name;
+    std::string_view proposed_cert = sm->get_outbound_cert();
+    Debug("http_ss", "validate_cert proposed_cert=%s, cert=%s", proposed_cert.data(), session_cert);
+    if (!session_cert || proposed_cert.length() == 0) {
+      retval = session_cert == nullptr && proposed_cert.length() == 0;
+    } else {
+      retval = proposed_cert.compare(session_cert) == 0;
+    }
+  }
+  return retval;
 }
 
 HSMresult_t
 ServerSessionPool::acquireSession(sockaddr const *addr, CryptoHash const &hostname_hash,
-                                  TSServerSessionSharingMatchType match_style, HttpSM *sm, Http1ServerSession *&to_return)
+                                  TSServerSessionSharingMatchMask match_style, HttpSM *sm, Http1ServerSession *&to_return)
 {
   HSMresult_t zret = HSM_NOT_FOUND;
   to_return        = nullptr;
 
-  if (TS_SERVER_SESSION_SHARING_MATCH_HOST == match_style) {
+  if ((TS_SERVER_SESSION_SHARING_MATCH_MASK_HOSTONLY & match_style) && !(TS_SERVER_SESSION_SHARING_MATCH_MASK_IP & match_style)) {
```

Comment:
You could think about using bitset for greater simplicity, with little or no performance impact, cause the optimizer is your daddy:  https://godbolt.org/z/wLvwum https://godbolt.org/z/zCphq4